Specialized allocation

jacob-hughes opened the pull request:
There are three ways a block can be allocated:

- Atomically - this means that a block contains no pointers which need
tracing during marking. An atomically allocated block is a leaf node in
the object graph.
 
- Precisely - blocks are allocated along with a bitmap which describes
where the pointers are on a per-word basis. The parts of a block which
are not described in the bitmap are ignored by the collector during
marking.
 
- Conservatively - the whereabouts of pointers inside the block are
unknown, and each word in the block must be considered for potential
pointers.

Users who enable the rustgc feature can now allocate memory with
`GcAllocator::maybe_optimised_alloc`. This method tries to allocate
blocks precisely (or even atomically) based on a value's type.  It first
tries to specialize allocation based on whether a type implements the
`Conservative` or `NoTrace` traits, and then uses the `gc_layout`
intrinsic in rustgc to try to allocate a block with a precise layout.

jacob-hughes committed:
Use single allocator

jacob-hughes committed:
Add specialized allocation to optimise marking

There are three ways a block can be allocated:

    Atomically - this means that a block contains no pointers which need
    tracing during marking. An atomically allocated block is a leaf node in
    the object graph.

    Precisely - blocks are allocated along with a bitmap which describes
    where the pointers are on a per-word basis. The parts of a block which
    are not described in the bitmap are ignored by the collector during
    marking.

    Conservatively - the whereabouts of pointers inside the block are
    unknown, and each word in the block must be considered for potential
    pointers.

Users who enable the rustgc feature can now allocate memory with
`GcAllocator::maybe_optimised_alloc`. This method tries to allocate
blocks precisely (or even atomically) based on a value's type.  It first
tries to specialize allocation based on whether a type implements the
`Conservative` or `NoTrace` traits, and then uses the `gc_layout`
intrinsic in rustgc to try to allocate a block with a precise layout.

jacob-hughes commented on src/spec.rs:
```diff
@@ -0,0 +1,179 @@
+///
+/// Specialized allocation can be thought of as "compile-time cascading if-else"
+/// which first tries to allocate atomically, then precisely, and then finally
+/// conservatively if neither of these optimised strategies are possible.
```
Ah, this sentence on ordering is misleading. I will remove it as I think it confuses matters.

jacob-hughes replied:
Fixed: d3fafa159230d25b3fe85b259eb36a6da215df5f

ltratt commented on src/spec.rs:
```diff
@@ -0,0 +1,176 @@
+//! The contents of this module require the rustgc feature. To avoid polluting
```
Maybe silly question: why is the file called `spec`?

jacob-hughes replied:
A I missed this one. It's just short for specialization/specializer. Perhaps its not a great name, but allocation_specialization felt a bit long winded.

ltratt replied:
Ah, I always think "spec" means "specification". I think we need a different filename, but I don't have a firm opinion on what that might be :)

jacob-hughes replied:
How about just `specialization`?

ltratt replied:
I'm fine with that. If you want something simpler, you could call it something like `picker` maybe?

jacob-hughes replied:
Let's go with `specialization`, I think if we went for `picker` we should rename the structs in the file for consistency.

ltratt replied:
I'm happy to go with whatever you decide on this one.

jacob-hughes replied:
Fixed 812827a440560dfd6d845cecd3953fe4009fac16

ltratt commented on src/spec.rs:
```diff
@@ -0,0 +1,176 @@
+impl<T> AtomicSpecializer for AllocationSpecializer<T> {
+    default fn maybe_optimised_alloc(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        let trace = unsafe { ::core::gc::gc_layout::<T>() };
+        if trace.is_conservative() {
```
What does `is_conservative` do? Is it defined in rustgc?

ltratt commented:
This is a surprisingly complex PR, and I've spent a while wrapping my head around what it's doing! A few comments.

First, there aren't any tests which makes me nervous. Is the intention that there'll be tests in an associated PR for `libgc`? [An aside: this PR made me wonder whether libgc/libgc_internal should be two crates in a single repo? Maybe that's not possible, or not a good idea: I don't remember if we'd considered it or not!]

Second, the specialisation strategy is implemented as a fall-through series of function calls where we hope (probably reasonably) that the optimiser will remove them. At the very least, this needs some sort of comment, because the flow is hard to discern. But I wondered if we even needed to do that? Is it possible to do something like:
```rust
impl <T> for AllocationSpecializer {
    default fn alloc(...);
}

impl <T: Conservative> for AllocationSpecializer {
    alloc(...);
}

impl <T: NoTrace> for AllocationSpecializer {
    alloc(...);
}
```
Perhaps it's the mysterious-to-me `is_conservative` call that makes this impossible, or something in Rust's specialisation that makes it impossible?

jacob-hughes commented:
> First, there aren't any tests which makes me nervous. Is the intention that there'll be tests in an associated PR for libgc?

Yes that's correct. It's really hard to add tests for the allocator functionality alone because you can't peek into the allocator after the fact and ask "which strategy were you allocated with?"

>  An aside: this PR made me wonder whether libgc/libgc_internal should be two crates in a single repo?

This wasn't chosen to make it easier to link with `rustgc`. When we build `rustgc`, we want to be able to pull in only `libgc_internal` so that we can access the allocator API. Rustc is setup uniquely which makes it difficult enough to pull external crates into the core library, but even more so if they are crates inside other crates. This separation was to help with that headache.

> Perhaps it's the mysterious-to-me is_conservative call that makes this impossible, or something in Rust's specialisation that makes it impossible?

It seems this is a limitation with Rust. It doesn't appear that you can specialize on more than one concrete case, hence the confusing and roundabout way I had to construct the control flow. The other thing to note is that `is_conservative` branches on the result of evaluation of a const expr (specifically, `core::intrinsics::gc_layout::<T>`). I then expect the optimiser to remove this branch since it should know the value at compile-time.

ltratt commented:
> It's really hard to add tests for the allocator functionality alone because you can't peek into the allocator after the fact and ask "which strategy were you allocated with?"

Tests don't *have* to do that (though obviously it's nice if they can!), but we do at least need to write tests which we believe tickle all the code paths. Without that, the chance of bit rot becomes orders of magnitude higher.

Is it the case that `is_conservative<T>` only returns `true` if `T` != `NoTrace` and == `Conservative`?

jacob-hughes commented:
> Is it the case that is_conservative<T> only returns true if T != NoTrace and == Conservative? 

I've just realised why there is confusion here. I have local changes to rustgc which I thought had been merged. I'll submit this as a PR now as it contains the `is_conservative` method and will explain what it does.

ltratt commented:
OK, I'll wait for that PR. I'd also like tests in this PR too please :)

jacob-hughes committed:
Split rustgc feature into two to allow unit testing

 The rustgc feature pulls in the "core" crate. This crate is a necessary
 hack so that libgc_internal can be included in rustc-std when rustgc
 depends on it.

Unfortunately, the "core" crate prevents libgc_internal from being
compiled in any other context - even cfg(test). Thus, in order to test
the code for the "rustgc" feature, we have to de-couple conditional
compilation from the dependencies that are added by the feature gate. As
of cargo 1.50, it's not possible to exclude dependencies based on
cfg(test). A hack to get around this is to use an additional feature
(rustgc_internal) to conditionally compile rustgc-specific code only.
This way, `cargo test --features rustgc_internal` can be used to build a
test binary for rustgc-specific features, while also being automatically
included as part of the `rustgc` feature when it is compiled as part of
rustgc-std.

jacob-hughes commented:
I've added some unit tests for this. Unfortunately, it required quite a few complex build changes to get them to work. I've tried to explain them as best I can in each commit. I won't lie though: it hurt my head! So feel free to ask questions.

jacob-hughes commented:
I've also renamed `is_conservative` to `must_use_conservative` here c03c04cd7d3d207369f67247759f99d0fdcc1c88. This requires https://github.com/softdevteam/rustgc/pull/22 to be merged first.

ltratt commented on src/spec.rs:
```diff
@@ -47,23 +50,33 @@ use crate::boehm;
 /// and Z contiguously (provided size and alignment constraints are met).
 /// Conservative allocation is currently the only safe way to allocate such
 /// blocks.
-pub(crate) struct AllocationSpecializer<T> {
```
A question: how were you able to get rid of the type parameter?

It's up to you but I wondered whether you might prefer something like:

```
#[cfg(not(test))] struct AllocationSpecializer;
#[cfg(test)] struct AllocationSpecializer { num_atomic: ..., ..., }
```

to make things clearer? It's perhaps as broad as it's long however!

jacob-hughes replied:
How about something like this: 18b60344b8cee4472d98dfec59f918f1670c57f6? 

> A question: how were you able to get rid of the type parameter?

It's still there, but instead of being constrained to the `AllocationSpecializer` struct, it exists on the `maybe_optimized_alloc::<T>` method. It doesn't make a huge difference really, especially as we construct a new `AllocationSpecializer` on each allocation (which is free, as it's a ZST). This does make the unit tests a bit tidier though because you can do multiple allocations of different types with the same `AllocationSpecializer`.

ltratt replied:
I like https://github.com/softdevteam/libgc_internal/commit/18b60344b8cee4472d98dfec59f918f1670c57f6: what are your thoughts on it though?

>This does make the unit tests a bit tidier though because you can do multiple allocations of different types with the same AllocationSpecializer.

Ah, makes sense.

jacob-hughes replied:
I think I prefer it. `cfg` attributes on each field is a bit dirty and obscures the fact that this should be a ZST in real-life. Shall we go with it?

ltratt replied:
Cool, let's keep it!

ltratt commented:
I feel a bit bad now because I didn't expect this to be so difficult to test! But it's probably to the good that we've bitten the bullet now rather than later :)

ltratt commented:
Please squash.

jacob-hughes commented:
Squashed, but we'll have to wait for https://github.com/softdevteam/rustgc/pull/22 before merging.

ltratt commented:
Agreed -- though that PR failed to build.

ltratt commented:
bors r+
